File: Arrays-Strings/power_set.cpp
#include<iostream>
#include<math.h>
#include<vector>
#include<limits>


using namespace std;


// Time Complexity: O(2^n)

// Approach: Either an element is selected or it is not selected.
// T(n)=2T(n-1)+c.


void power_set(int input[], int n, vector<int> curr, int **ps, int *i);


int main() {
	int n;
	cout << "Enter the number of numbers you want in array: ";
	cin >> n;
	int arr[n];
	cout << "Enter the numbers: " << endl;
	for(int i=0;i<n;i++) {
		cin >> arr[i];
	}
	// Vector used as a stack for inserting numbers.
	vector<int> curr;
	int x=0;
	//Used for filling row by row of the resultant array.
	int *k=&x;
	//Resultant Array.
	int **ps=(int **)malloc(((int)pow(2,n))*sizeof(int *));
	for(int i=0;i<(int)pow(2,n);i++) {
		*(ps+i)=(int *) malloc((n+1)*sizeof(int));
	}
	power_set(arr,n,curr,ps,k);
	cout << "The Power Set: "<<endl;
	cout << "0" <<endl;
	//Printing the resultant array.
	for(int i=0;i<(int)pow(2,n);i++) {
		for(int j=0;j<n;j++) {
			if(ps[i][j]==INT_MAX) {
				break;
			}
			cout << ps[i][j] << "\t";
		}
		cout << endl;
	}
}

void power_set(int input[], int n, vector<int> curr, int **ps, int *i) {
	// If n=0, then add the vector to i row of the resultant array.
	if(n==0) {
		vector<int>::iterator x;
		int j=0;
		for(j=0;j<curr.size();j++) {
			ps[*i][j]=curr[j];
		}
		ps[*i][j]=INT_MAX;
		// Row increased for each array.
		(*i)+=1;
		return;
	}

	//Element Selected.
	curr.push_back(input[n-1]);


	power_set(input,n-1,curr,ps,i);

	//Element Not Selected
	curr.pop_back();

	power_set(input,n-1,curr,ps,i);


}

File: Arrays-Strings/PowerSet.java
import java.util.Scanner;

// Time Complexity: O(n*(2^n)).
// Approach Bit Masking. 
// 1<<j this is a left shift bitwise operator. It will shift bits of 1 to left by j. For eg: 1<<3=> 1000=>8. 
// 1<<j also means multiplying 1 with 2^j.
// 1>>j means dividing 1 with 2^j.
// Then, this value will get AND with the bits of i.
// If there is a 1 at same positions in the bits of i as well as (1<<j), then that bit will remain 1 after AND operation.
// Then, the result will be greater than 1.
// This is called bit setting/masking.

public class PowerSet {

  private static final Scanner sc = new Scanner(System.in);

  public static void main(String[] args) {
    int n;
    System.out.println("Enter the number of elements:");
    n = sc.nextInt();
    char[] arr = new char[n];
    System.out.println("Enter the elements: ");
    //Taking input.
    for (int i = 0; i < n; i++) {
      arr[i] = sc.next().charAt(0);
    }
    printPowerSet(arr, n);
  }

  private static void printPowerSet(char[] arr, int n) {
    int count = (int) Math.pow(2, n);
    // For each no in the power set i.e. in for each no: i in 2^n elements.
    for (int i = 0; i < count; i++) {
      // For each number j.
      for (int j = 0; j < n; j++) {
        // setting the bits.
        if ((i & (1 << j)) > 0) {
          System.out.print(arr[j]);
        }
      }
      System.out.println();
    }
  }
}